Source/FieldSolver/FiniteDifferenceSolver/MacroscopicEvolveHM_2nd.cpp

```diff
@@ -0,0 +1,902 @@
+                            Hz_eff += MacroscopicProperties::face_avg_to_face(i, j, k, 0, amrex::IntVect(0, 0, 1), amrex::IntVect(0, 0, 1), Hz);
+                        }
+
+                        // calculate the a_temp_static_coeff (it is divided by 2.0 because the input dt is actually dt/2.0)
+                        Real a_temp_dynamic_coeff = PhysConst::mu0 * amrex::Math::abs(mag_gamma_arrz) / 2.0;
+
+                        // calculate a_temp_zface
+                        // x component on z-faces of grid
+                        a_temp_zface(i, j, k, 0) = (M_normalization != 0) ? -(dt * a_temp_dynamic_coeff * Hx_eff + a_temp_static_zface(i, j, k, 0))
+                                                                          : -(dt * a_temp_dynamic_coeff * Hx_eff + 0.5 * a_temp_static_zface(i, j, k, 0) 
+                                                                            + 0.5 * mag_alpha_arrz * 1. / std::sqrt(std::pow(M_zface(i, j, k, 0), 2.0) + std::pow(M_zface(i, j, k, 1), 2.0) + std::pow(M_zface(i, j, k, 2), 2.0)) * M_old_zface(i, j, k, 0));
+
+                        // y component on z-faces of grid
+                        a_temp_zface(i, j, k, 1) = (M_normalization != 0) ? -(dt * a_temp_dynamic_coeff * Hy_eff + a_temp_static_zface(i, j, k, 1))
+                                                                          : -(dt * a_temp_dynamic_coeff * Hy_eff + 0.5 * a_temp_static_zface(i, j, k, 1) 
+                                                                            + 0.5 * mag_alpha_arrz * 1. / std::sqrt(std::pow(M_zface(i, j, k, 0), 2.0) + std::pow(M_zface(i, j, k, 1), 2.0) + std::pow(M_zface(i, j, k, 2), 2.0)) * M_old_zface(i, j, k, 1));
+
+                        // z component on z-faces of grid
+                        a_temp_zface(i, j, k, 2) = (M_normalization != 0) ? -(dt * a_temp_dynamic_coeff * Hz_eff + a_temp_static_zface(i, j, k, 2))
+                                                                          : -(dt * a_temp_dynamic_coeff * Hz_eff + 0.5 * a_temp_static_zface(i, j, k, 2) 
+                                                                            + 0.5 * mag_alpha_arrz * 1. / std::sqrt(std::pow(M_zface(i, j, k, 0), 2.0) + std::pow(M_zface(i, j, k, 1), 2.0) + std::pow(M_zface(i, j, k, 2), 2.0)) * M_old_zface(i, j, k, 2));
+
+                        // update M_zface from a and b using the updateM_field
+                        // x component on z-faces of grid
+                        M_zface(i, j, k, 0) = MacroscopicProperties::updateM_field(i, j, k, 0, a_temp_zface, b_temp_static_zface);
+
+                        // y component on z-faces of grid
+                        M_zface(i, j, k, 1) = MacroscopicProperties::updateM_field(i, j, k, 1, a_temp_zface, b_temp_static_zface);
+
+                        // z component on z-faces of grid
+                        M_zface(i, j, k, 2) = MacroscopicProperties::updateM_field(i, j, k, 2, a_temp_zface, b_temp_static_zface);
```

Comment:
Can you please put this in a for  loop as was done for x and yfaces above?

Reply:
Don't - see comment above (Hx_eff, Hy_eff, Hz_eff)

Reply:
We can create an auxiliary H_eff(i,j,k,icomp), which is a vector of vectors containing Hx_eff, Hy_eff and Hz_eff?

Reply:
Good idea - makes the code way shorter.